contract/src/utils.rs

```diff
@@ -30,14 +35,14 @@ pub(crate) fn upsert_dictionary_value_from_key<T: CLTyped + FromBytes + ToBytes>
         NFTCoreError::InvalidStorageUref,
     );
 
-    // TODO: Write a test for this upsert method.
     match storage::dictionary_get::<T>(seed_uref, key) {
         Ok(None | Some(_)) => storage::dictionary_put(seed_uref, key, value),
         Err(error) => runtime::revert(error),
     }
 }
 
```

Comment:
nitpick: IMO, it feels like we are stretching the term "utils" a bit far here. it's mostly the modality enums and some adjacent functions and rather than being utilitarian in nature the modality enums are essential to the definition and logic of the smart contract proper. 

it might make sense to rename the file "modes.rs" or "modalities.rs"

the metadata schema structs and supporting logic could be moved off to a file called "metadata.rs", or potentially a subfolder with each of the metadata schema structs in its own file for ease of discovery.

Reply:
Agreed, would like to mark that as follow up work in a separate PR